packages/jupyter-ai/src/inline-completions.ts

```diff
@@ -0,0 +1,348 @@
+          {
+            label: 'Show Traceback',
+            callback: () => {
+              showErrorMessage(
+                'Inline completion failed on the server side',
+                error.traceback
+              );
+            }
+          }
+        ]
+      });
+      throw new Error(
+        `Inline completion failed: ${error.type}\n${error.traceback}`
+      );
+    }
+    return result.list;
+  }
+
+  get schema(): ISettingRegistry.IProperty {
+    return {
+      properties: {
+        maxPrefix: {
+          title: 'Maximum prefix length',
+          minimum: 1,
+          type: 'number',
+          description:
+            'At most how many prefix characters should be provided to the model.'
+        },
+        maxSuffix: {
+          title: 'Maximum suffix length',
+          minimum: 0,
+          type: 'number',
+          description:
+            'At most how many suffix characters should be provided to the model.'
+        }
+      },
+      default: JupyterAIInlineProvider.DEFAULT_SETTINGS as any
+    };
+  }
+
+  async configure(settings: { [property: string]: JSONValue }): Promise<void> {
+    this._settings = settings as unknown as JupyterAIInlineProvider.ISettings;
+  }
+
+  /**
+   * Extract prefix from request, accounting for context window limit.
+   */
+  private _prefixFromRequest(request: CompletionHandler.IRequest): string {
+    const textBefore = request.text.slice(0, request.offset);
+    const prefix = textBefore.slice(
+      -Math.min(this._settings.maxPrefix, textBefore.length)
+    );
+    return prefix;
+  }
+
+  /**
+   * Extract suffix from request, accounting for context window limit.
+   */
+  private _suffixFromRequest(request: CompletionHandler.IRequest): string {
+    const textAfter = request.text.slice(request.offset);
+    const prefix = textAfter.slice(
+      0,
+      Math.min(this._settings.maxPrefix, textAfter.length)
+    );
+    return prefix;
+  }
+
+  private _resolveLanguage(language: IEditorLanguage | null) {
+    if (!language) {
+      return 'plain English';
+    }
+    if (language.name === 'ipythongfm') {
+      return 'markdown';
+    }
+    return language.name;
+  }
+
+  private _settings: JupyterAIInlineProvider.ISettings =
+    JupyterAIInlineProvider.DEFAULT_SETTINGS;
+
+  private _currentModel = '';
+  private _counter = 0;
+}
+
+namespace JupyterAIInlineProvider {
+  export interface IOptions {
+    completionHandler: CompletionWebsocketHandler;
+    languageRegistry: IEditorLanguageRegistry;
+  }
+  export interface ISettings {
+    maxPrefix: number;
+    maxSuffix: number;
+    debouncerDelay: number;
+  }
+  export const DEFAULT_SETTINGS: ISettings = {
+    maxPrefix: 10000,
+    maxSuffix: 10000,
+    // The debouncer delay handling is implemented upstream in JupyterLab;
+    // here we just increase the default from 0, as compared to kernel history
+    // the external AI models may have a token cost associated.
+    debouncerDelay: 250
```

Comment:
I am not 100% sure if we want to make it opt-in or opt-out but I _think_ it should be opt-out at least in the first release. Do you agree?

```suggestion
    debouncerDelay: 250,
    enabled: false
```